### PR TITLE
Allow DynamicFileCatalog support to query partitioned file

### LINKED ISSUE #12671: Allow DynamicFileCatalog support to query partitioned file
### Is your feature request related to a problem or challenge?

#11035 supports to query files through their URL. If the target dataset is partitioned, `DynamicFileCatalog` can't recognize the partitioned columns well. 
Given the file structure like:
```
partitioned_table
     c_date=2018-11-13
         data.csv 
     c_date=2018-12-13
         data.csv
```
If we tried to query it through the dynamic file catalog
```rust
    let sql = "SELECT * FROM 'datafusion/core/tests/data/partitioned_table'";
    let session_config =
        SessionConfig::new().set_str("datafusion.catalog.has_header", "false");
    let state = SessionStateBuilder::default()
        .with_default_features()
        .with_config(session_config)
        .build();
    let ctx = SessionContext::new_with_state(state).enable_url_table();
    ctx.sql(sql).await?.show().await?;
```
The result is
```
+----------+-------------------------+
| column_1 | column_2                |
+----------+-------------------------+
| Jorge    | 2018-12-13T12:12:10.011 |
| Andrew   | 2018-11-13T17:11:10.011 |
| Jorge    | 2018-12-13T12:12:10.011 |
| Andrew   | 2018-11-13T17:11:10.011 |
+----------+-------------------------+
```

The partitioned column `c_date` won't be used.

### Describe the solution you'd like

When inferring the `ListingTableConfig`, we can register the table partition column automatically. I think we can invoke `ListingOption::infer_partitions` to infer the required partition columns at the runtime.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #12671.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/dynamic_file.slt [hunk 1]
```diff
@@ -25,9 +25,170 @@ SELECT * FROM '../core/tests/data/partitioned_table_arrow/part=123' ORDER BY f0;
 1 foo true
 2 bar false
 
-# dynamic file query doesn't support partitioned table
-statement error DataFusion error: Error during planning: table 'datafusion.public.../core/tests/data/partitioned_table_arrow' not found
-SELECT * FROM '../core/tests/data/partitioned_table_arrow' ORDER BY f0;
```
Comment (goldmedal) [inline datafusion/sqllogictest/test_files/dynamic_file.slt:30]: It's what I mentioned in https://github.com/apache/datafusion/issues/12671#issuecomment-2381291431
I think it isn't a dynamic file catalog issue, so I removed it from the test.

File: datafusion/core/src/datasource/listing/table.rs [hunk 2]
```diff
@@ -167,6 +187,14 @@ impl ListingTableConfig {
         })
     }
 
+    fn is_enable_url_table(session: &SessionState) -> bool {
+        session
+            .catalog_list()
+            .as_any()
+            .downcast_ref::<DynamicFileCatalog>()
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:194]: The only thing I am concerned about here is that this is a special case for DynamicFileCatalog which is tightly coupling the catalog to the ListingTable (when the hope is that we can eventually split them apart)

Would it be possible instead to add an explicit option to `ListingTableConfig` like `infer_partitions_from_path`

And then set that option when creating the `ListingTableConfig` in `DynamicTableProvider`?

That way this code will still work but we won't have a direct dependency on DynamicTableProvider

**Review by alamb — COMMENTED** [review]
Thank you @goldmedal  -- this looks very close. I think we just need to avoid the back dependency on catalog and this PR will be good to go

**Review by alamb — APPROVED** [review]
Looks good to me @goldmedal  -- thank you for the contribution 🦾

**Comment by alamb** [comment]
🚀

**Pull request merged**